### ISSUE TITLE
Bokeh serve hanging in between client-server connection on add_root() 

### ISSUE THREAD

**Issue by peteawest** [bokeh/bokeh#4071]
Hi Bokeh, 
When running a the following command "bokeh serve --show server1.py" my add_root() method seems to hang in between client-server communication. I've attempted not using the "bokeh serve" command and instead incorporating "session = push_session(curdoc()) ... session.loop_until_closed() " but that hasn't produced a solution. When running with "bokeh serve...", the client socket connection does not close and the server seems to hang.

I'd like to be able to have the bokeh server running and constantly be creating client socket connections that send, update/add to the existing data, and close. 

@birdsarah had a great response to @dwf in issue #3087 and that has helped me out so  #far...

Here's the bokeh serve command that I execute first "bokeh serve --show server1.py":

```
from socket import *
from time import ctime

HOST = 'localhost'
PORT = 5006
BUFSIZ = 2048
ADDR = (HOST, PORT)

tcpSerSock = socket(AF_INET, SOCK_STREAM)
tcpSerSock.bind(ADDR)
tcpSerSock.listen(5)

import numpy as np

from bokeh.palettes import RdYlBu3
from bokeh.plotting import figure, curdoc, vplot

p = figure(x_range=(0, 100), y_range=(0, 100), toolbar_location=None)
p.border_fill_color = 'black'
p.background_fill_color = 'black'
p.outline_line_color = None
p.grid.grid_line_color = None

r = p.text(x=[], y=[], text=[], text_color=[], text_font_size="20pt",
           text_baseline="middle", text_align="center")

ds = r.data_source

def update(i):
    i = int(i)
    ds.data['x'].append(np.random.random()*70 + 15)
    ds.data['y'].append(np.random.random()*70 + 15)
    ds.data['text_color'].append(RdYlBu3[i%3])
    ds.data['text'].append(str(i))
    ds.trigger('data', ds.data, ds.data)

while True:
    print('waiting for connection...')
    tcpCliSock, addr = tcpSerSock.accept()
    print('connected from: ', addr)

    while True:
        data = tcpCliSock.recv(BUFSIZ)
        if not data:
            break
        rdata = data.decode('utf-8')
        response = '[%s] %s' % (ctime(), rdata)

        curdoc().add_root(update(rdata), p)
        print("rdata:", rdata)
        print("ds:", ds.data['x'], ds.data['y'])
        print("p:", p)
  
        tcpCliSock.send(response.encode('utf-8'))
        print(response)

    tcpCliSock.close()
tcpSerSock.close()

```

Here's the client command that I run second to send data "python client1.py" :
```
from socket import *

HOST = 'localhost'
PORT = 5006
BUFSIZ = 2048
ADDR = (HOST, PORT)

tcpCliSock = socket(AF_INET, SOCK_STREAM)
tcpCliSock.connect(ADDR)

data = '5'
tcpCliSock.send(data.encode('utf-8'))
data = tcpCliSock.recv(BUFSIZ)
print(data.decode('utf-8'))

tcpCliSock.close()
```

When at the "localhost..." url, the figure loads and I see what I want. However, the server connection hangs and won't close. Is this an issue or my coding error? Please redirect to the mailing list if this is not an issue and just a misunderstanding of my introductory knowledge of bokeh serve.

Thanks for your time and consideration in help!

**Comment by bryevdv** [comment]
@peteawest I am traveling this week so I apologize I cannot be more detailed at the moment. The problem is almost surely the `while True`. A Bokeh App needs to *return*. So a loop like that is not going to work. Other folks have successfully gotten outside asynchronous events to trigger callbacks that update a Bokeh plot, you might look to the extended discussion in #4000 especially re: `next_tick_callback`

**Comment by bryevdv** [comment]
ping @havocp in case you have time to chime in

**Comment by havocp** [comment]
There are essentially two options to do networking in your app. One is to use the Tornado APIs such as their async http client or io streams. Two is to use a thread but the thread cannot touch Bokeh or Tornado directly, all it can do is add a next tick callback or perhaps complete a future. #4000 discusses possible current caveats with threads. The thread should not touch Bokeh documents or models.

**Comment by havocp** [comment]
I should add a third option which is to do plain old blocking IO and accept that the app will lock up during it. For many apps this is fine, for example you could do a periodic callback that loads new data every 10 seconds, and if loading takes 1 second the app is just locked up for 1 second every 10.  Or you could load data when the user clicks a button or something. For apps with only one or two users at a time this is probably fine, for multiuser apps it probably isn't.

**Comment by peteawest** [comment]
Thank you @havocp and @bryevdv for your helpful responses. I'm new to networking/threading so creating a networking bokeh app is challenging from the get go. I was able to achieve a solution using @havocp suggestion "... use a thread but the thread cannot touch Bokeh or Tornado directly, all it can do is add a next tick callback...". I think the best solution would be one that implements the Tornado APIs, but that's a steep learning curve for me at the moment. I hope to investigate that soon. 

I run the below code as "python server1.py" and then run "python client1.py" (mentioned above with port 9999) from a second terminal. It's slow and not ideal. Maybe some one can chime in with a way to improve, but you guys have already been very helpful! Thanks and I hope to keep working on this.

```
import threading
import time
from socket import *
from time import ctime
from bokeh.client import push_session
from bokeh.plotting import figure, curdoc

worker_exit_flag = 0
thread_exit_flag = 0

def worker():
    global i
    i = "0,0"
    HOST = 'localhost'
    PORT = 9999
    BUFSIZ = 1024
    ADDR = (HOST, PORT)
    tcpSerSock = socket(AF_INET, SOCK_STREAM)
    tcpSerSock.bind(ADDR)
    tcpSerSock.listen(5)
    while worker_exit_flag == 0:
        print('waiting for connection...')
        tcpCliSock, addr = tcpSerSock.accept()
        print('connected from: ', addr)
        data = tcpCliSock.recv(BUFSIZ)
        if not data:
            break
        i = data.decode('utf-8')
        response = '[%s] %s' % (ctime(), i)
        tcpCliSock.send(response.encode('utf-8'))
        print(response)
        tcpCliSock.close()
    tcpSerSock.close()
    print('Exiting worker')

def my_thread():
    global x
    global y
    x = 0
    y = 0
    while thread_exit_flag == 0:
        d = i.split(',')
        x = int(d[0])
        y = int(d[1])
        time.sleep(1)
    print('Exiting thread')

if __name__ == '__main__':
    w = threading.Thread(name='worker', target=worker)
    t = threading.Thread(name='thread', target=my_thread)
    w.start()
    t.start()
    p = figure()
    c = p.circle(x=[0],y=[0])
    ds = c.data_source
    def update():
        ds.data['x'] = [x]
        ds.data['y'] = [y]
        print(ds.data['x'])
    curdoc().add_periodic_callback(update, 500)
    session = push_session(curdoc())
    session.show()
    session.loop_until_closed()
```

**Comment by bryevdv** [comment]
A few notes: 

* You should always update `.data` in one go, i.e. make a new dict and set `.data` to it, don't update `'x'` and `'y'` separately as you have done above. 

* If you are streaming in data to append, you should look at the `.stream` API show in `examples/app/ohlc`

* I think we made the tornado suggestion seem more complicated than it is. It really just boils down to doing something like:
```
curdoc().add_next_tick_callback(update)
``` 
whenever there is new data (instead of the periodic callback). The point is that `add_next_tick_callback` is the only thread-safe way to interact with the event loop.

**Comment by bryevdv** [comment]
It turns out there is probably an issue that needs resolving with `next_tick_callback` I can't guarantee that it will be helpful until it is resolved (it depends), see #3072 for details

**Comment by bryevdv** [comment]
The needed features have since been implemented. See:

http://bokeh.pydata.org/en/latest/docs/user_guide/server.html#updating-from-threads

and 

http://bokeh.pydata.org/en/latest/docs/user_guide/server.html#updating-from-unlocked-callbacks
